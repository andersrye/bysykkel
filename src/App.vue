<template>
  <div id="app">
    <StationMap
      class="bicycle-map"
      :stations="stations"
      :selected-station="selectedStation"
      :show-bikes="showBikes"
    />
    <button
      class="lower-left-button button is-light"
      v-if="showBikes"
      @click="showBikes = false"
    >
      <span class="icon">
        <FontAwesomeIcon icon="lock-open" />
      </span>
      <span>Vis ledige plasser</span>
    </button>
    <button
      class="lower-left-button button is-link"
      v-if="!showBikes"
      @click="showBikes = true"
    >
      <span class="icon">
        <FontAwesomeIcon icon="bicycle" />
      </span>
      <span>Vis ledige sykler</span>
    </button>

    <transition name="fade">
      <button
        class="lower-right-button button is-link"
        v-if="!showSearch"
        @click="showSearch = true"
      >
        <span class="icon">
          <FontAwesomeIcon icon="search" />
        </span>
        <span>SÃ¸k</span>
      </button>
    </transition>
    <transition name="fade">
      <StationSearch
        v-if="showSearch"
        class="station-search"
        :stations="stations"
        @close="showSearch = false"
        @selected-station="stationSelected"
      />
    </transition>
  </div>
  <span
    class="tag is-warning status"
    v-if="status === 'loading'"
  >
    <span class="icon"><FontAwesomeIcon
      icon="spinner"
      class="fa-spin"
    /></span>
    <span>Loading</span>
  </span>
  <span
    class="tag is-danger status"
    v-if="status === 'error'"
  >
    <span class="icon"><FontAwesomeIcon icon="spinner"/></span>
    <span>Loading</span>
  </span>
  <span
    class="tag is-success status"
    v-if="status === 'ok'"
  ><span class="icon"><FontAwesomeIcon icon="circle-check"/></span></span>
</template>

<script setup>
import {ref, computed, watchEffect} from "vue"
import StationMap from "./components/StationMap"
import StationSearch from "./components/StationSearch"
import Gbfs from './gbfs'
import sources from './gbfs-sources.json'
import {FontAwesomeIcon} from '@fortawesome/vue-fontawesome'

const systemInfo = ref(null)
const stationInfo = ref(null)
const stationStatus = ref(null)
const selectedStation = ref(null)
const showBikes = ref(true)
const showSearch = ref(false)
const status = ref("loading")

const stations = computed(() => {
  //merge the station info and status to make it easier to handle
  const statusById = stationStatus.value?.data?.stations?.reduce((acc, status) => {
    acc[status.station_id] = status
    return acc
  }, {}) ?? {}
  return stationInfo.value?.data?.stations.map(info => ({...info, ...statusById[info.station_id]}))
})

const source = new URLSearchParams(window.location.search).get('source') ?? 'oslo'
const gbfs = new Gbfs(sources[source])

gbfs.getSystemInfo()
    .then(res => systemInfo.value = res)
    .catch(handleError)
gbfs.getStationInfo()
    .then(res => stationInfo.value = res)
    .catch(handleError)
pollStationStatus()
    .catch(handleError)

watchEffect(() => {
  if(systemInfo.value && stationInfo.value && stationStatus.value) {
    status.value = 'ok'
  }
})

async function pollStationStatus() {
  for await (const status of gbfs.streamStationStatus()) {
    stationStatus.value = status
  }
}

function stationSelected(station) {
  showSearch.value = false
  selectedStation.value = station
}

function handleError(error) {
  console.error(error)
  //todo: this is not a great way to handle errors. automatically retry and update status
  status.value = 'error'
}
</script>

<style scoped>

.bicycle-map {
  position: absolute;
  overflow: hidden;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.station-search {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  margin: 8px 8px 32px 8px;
  height: calc(100vh - 40px);
  width: calc(100vw - 32px);
  max-width: 500px;
  overflow: auto;
}

.lower-left-button {
  position: absolute;
  left: 0;
  bottom: 0;
  margin: 8px 8px 36px 8px;
}

.lower-right-button {
  position: absolute;
  right: 0;
  bottom: 0;
  margin: 8px 8px 36px 8px;
}

.status {
  position: absolute;
  top: 0;
  left: 0;
  margin: 8px
}

.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.3s ease;
}

.fade-enter-from,
.fade-leave-to {
  opacity: 0;
}
</style>
