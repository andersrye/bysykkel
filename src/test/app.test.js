import {flushPromises, shallowMount} from '@vue/test-utils'
import App from '@/App.vue'
import { expect, test, vi } from 'vitest'
import Gbfs from "@/gbfs";

const mockGenerator = function*(yieldValue) { yield yieldValue }

vi.mock("@/gbfs", () => {
    const Gbfs = vi.fn()
    Object.assign(Gbfs.prototype, {
        getSystemInfo: vi.fn().mockResolvedValue({}),
        getStationInfo: vi.fn().mockResolvedValue({}),
        getStationStatus: vi.fn().mockResolvedValue({}),
        streamStationStatus: vi.fn(() => mockGenerator({}))
    })
    return { default: Gbfs }
})

describe('app', () => {
    let gbfs

    beforeEach(() => {
        gbfs = new Gbfs("")
    })

    afterEach(() => {
        vi.clearAllMocks()
    })

    test('mount component', async () => {
        const wrapper = shallowMount(App)
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('update status on success', async () => {
        const wrapper = shallowMount(App)

        expect(wrapper.get('[data-test=status-indicator]').attributes('status')).toBe('loading')
        await flushPromises() //wait for data to "load"
        expect(wrapper.get('[data-test=status-indicator]').attributes('status')).toBe('ok')
    })

    test('update status on error', async () => {
        gbfs.getSystemInfo.mockRejectedValueOnce("Mock failure")
        const wrapper = shallowMount(App)

        expect(wrapper.get('[data-test=status-indicator]').attributes('status')).toBe('loading')
        await flushPromises()
        expect(wrapper.get('[data-test=status-indicator]').attributes('status')).toBe('error')
    })

    test('bike buttons should toggle between bikes and docks', async () => {
        const wrapper = shallowMount(App)

        expect(wrapper.find('[data-test=bikes-button]').exists()).toBe(false)
        expect(wrapper.find('[data-test=docks-button]').exists()).toBe(true)
        expect(wrapper.get('[data-test=map]').attributes('showbikes')).toBe('true')

        await wrapper.get('[data-test=docks-button]').trigger('click')

        expect(wrapper.find('[data-test=bikes-button]').exists()).toBe(true)
        expect(wrapper.find('[data-test=docks-button]').exists()).toBe(false)
        expect(wrapper.get('[data-test=map]').attributes('showbikes')).toBe('false')

    })

    test('search buttons should show search view', async () => {
        const wrapper = shallowMount(App)

        expect(wrapper.find('[data-test=search-view]').exists()).toBe(false)

        await wrapper.get('[data-test=search-button]').trigger('click')

        expect(wrapper.find('[data-test=search-view]').exists()).toBe(true)

    })

    //todo: test computed combined station info
})
