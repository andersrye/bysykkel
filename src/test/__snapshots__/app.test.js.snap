// Vitest Snapshot v1

exports[`app > mount component 1`] = `
"<div id=\\"app\\" data-v-7ba5bd90=\\"\\">
  <station-map-stub stations=\\"\\" showbikes=\\"true\\" bounds=\\"10.81431433498766,59.9534114,10.651118,59.90321254732584\\" class=\\"bicycle-map\\" data-test=\\"map\\" data-v-7ba5bd90=\\"\\"></station-map-stub><button class=\\"lower-left-button button is-light\\" data-test=\\"docks-button\\" data-v-7ba5bd90=\\"\\"><span class=\\"icon\\" data-v-7ba5bd90=\\"\\"><font-awesome-icon-stub icon=\\"lock-open\\" border=\\"false\\" fixedwidth=\\"false\\" flip=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" bounce=\\"false\\" shake=\\"false\\" beat=\\"false\\" fade=\\"false\\" beatfade=\\"false\\" flash=\\"false\\" spinpulse=\\"false\\" spinreverse=\\"false\\" data-v-7ba5bd90=\\"\\"></font-awesome-icon-stub></span><span data-v-7ba5bd90=\\"\\">Vis ledige plasser</span></button>
  <!--v-if-->
  <status-indicator-stub status=\\"loading\\" class=\\"status-indicator\\" data-test=\\"status-indicator\\" data-v-7ba5bd90=\\"\\"></status-indicator-stub>
  <transition-stub data-v-7ba5bd90=\\"\\"><button class=\\"lower-right-button button is-link\\" data-test=\\"search-button\\" data-v-7ba5bd90=\\"\\"><span class=\\"icon\\" data-v-7ba5bd90=\\"\\"><font-awesome-icon-stub icon=\\"search\\" border=\\"false\\" fixedwidth=\\"false\\" flip=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" bounce=\\"false\\" shake=\\"false\\" beat=\\"false\\" fade=\\"false\\" beatfade=\\"false\\" flash=\\"false\\" spinpulse=\\"false\\" spinreverse=\\"false\\" data-v-7ba5bd90=\\"\\"></font-awesome-icon-stub></span><span data-v-7ba5bd90=\\"\\">SÃ¸k</span></button></transition-stub>
  <transition-stub data-v-7ba5bd90=\\"\\">
    <!--v-if-->
  </transition-stub>
</div>"
`;
